cmd/util/ledger/migrations/add_key_migration.go

```diff
@@ -0,0 +1,391 @@
+package migrations
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/cadence/common"
+	"github.com/onflow/crypto/hash"
+
+	"github.com/onflow/flow-go-sdk/crypto"
+	"github.com/onflow/flow-go/cmd/util/ledger/reporters"
+	"github.com/onflow/flow-go/cmd/util/ledger/util/registers"
+	"github.com/onflow/flow-go/fvm"
+	"github.com/onflow/flow-go/fvm/systemcontracts"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// This migration is not safe to run on actual networks.
+// It is used for getting control over system accounts so that they can be tested
+// in a copied environment. When we need to do this it is best to create a branch
+// where this variable is set to true, and use that temporary branc to run migrations.
+const IAmSureIWantToRunThisMigration = false
+
+// AddKeyMigration adds a new key to the core contracts accounts
+type AddKeyMigration struct {
+	log zerolog.Logger
+
+	accountsToAddKeyTo map[common.Address]AddKeyMigrationAccountPublicKeyData
+	chainID            flow.ChainID
+
+	reporter reporters.ReportWriter
+}
+
+var _ AccountBasedMigration = &AddKeyMigration{}
+
+func NewAddKeyMigration(
+	chainID flow.ChainID,
+	key crypto.PublicKey,
+	rwf reporters.ReportWriterFactory,
+) *AddKeyMigration {
+	if !IAmSureIWantToRunThisMigration {
+		panic("Cannot run AddKeyMigration migration")
+	}
+
+	addresses := make(map[common.Address]AddKeyMigrationAccountPublicKeyData)
+	sc := systemcontracts.SystemContractsForChain(chainID).All()
+
+	for _, sc := range sc {
+		addresses[common.Address(sc.Address)] = AddKeyMigrationAccountPublicKeyData{
+			PublicKey: key,
+			HashAlgo:  hash.SHA3_256,
+		}
+	}
+
+	// add key to node accounts
+	for _, nodeAddress := range mainnetNodeAddresses {
```

Comment:
As double protection, I would suggest to use a `emptyNodeAddresses` here. So in order to use it, both `emptyNodeAddresses` and `IAmSureIWantToRunThisMigration` have to be updated. Updating only one would only be a no-op.

```suggestion
	for _, nodeAddress := range emptyNodeAddresses {
```

Reply:
Do you think these will change over time? wouldn't it be better to just have them prepared here?

Reply:
I mean we keep the `mainnetNodeAddresses` variable, but we create another `emptyNodeAddresses` variable. And by default the `emptyNodeAddresses` is used, so that this is a double protect. 

Then, in order to use it in migration mainnet, we need to do 2 things:
1. set `IAmSureIWantToRunThisMigration` to true
2. replace `emptyNodeAddresses` with `mainnetNodeAddresses` here. 

And if we forgot to do step 2, the `IHaveReplacedEmptyNodeAddressesToMainnetAddresses` is still false, and the error will remind us. 

And if we updated `IHaveReplacedEmptyNodeAddressesToMainnetAddresses` to true, but didn't actually do step 2, then it will still be a no-op. 

Does it make sense?

Reply:
what about requiring a specific chainID instead? 

I think it's more likely that we accidentally merge code with the updated config so adding a check that we wouldn't need to change would help avoid accidental merges allowing this migration against a main network.